File: lib/cpp/rpc/Transport.cpp
#include "rpc/Transport.h"
#include "rpc/NetEvent.h"
#include "rpc/Connection.h"

const size_t MsgHeaderMaxSize=2;
void rpc::ITransport::conn_eventcb(bufferevent * bev, short events, void * userData) {
	ITransport* trans = (ITransport*)userData;
	trans->conn_->handleClose();
}

void rpc::ITransport::conn_readcb(bufferevent * bev, void * userData) {
	ITransport* trans = (ITransport*)userData;
	struct evbuffer *input = bufferevent_get_input(bev);

	trans->conn_->handleInput(input);
}
void rpc::ITransport::conn_writecb(bufferevent * bev, void * userData) {
	ITransport* trans = (ITransport*)userData;
}

int rpc::TcpTransport::sendRequestMsg( const RequestMsg& reqMsg) {
	BinaryProtocol proto;
	reqMsg.serialize(&proto);

	uint16 len = uint16(proto.getBuffer().size());
	struct evbuffer * msgbuff= evbuffer_new();
	evbuffer_add(msgbuff,(byte*)&len,MsgHeaderMaxSize);
	evbuffer_add(msgbuff,proto.getBuffer().data(), proto.getBuffer().size());

	int ret = bufferevent_write_buffer(bev_, msgbuff);

	evbuffer_free(msgbuff);

	return ret;
}

int rpc::TcpTransport::recvResponseMsg(struct evbuffer* buff,std::vector<ResponseMsg>& msgVec) {
	int ret=0;
	while(1){
		if (evbuffer_get_length(buff) <= MsgHeaderMaxSize) {
			return ret;
		}
		struct evbuffer_iovec v[1];
		ret=evbuffer_peek(buff, MsgHeaderMaxSize, NULL, v, 1);
		RpcAssert(ret==1);
		uint16 msgLen= *(uint16*)(v[0].iov_base);

		if (evbuffer_get_length(buff) < size_t(msgLen+MsgHeaderMaxSize)) {
			return ret;
		}
		evbuffer_drain(buff,MsgHeaderMaxSize);
		ret+=MsgHeaderMaxSize;

		std::unique_ptr<byte[]> buffArray = std::make_unique<byte[]>(msgLen);
		evbuffer_remove(buff, buffArray.get(), msgLen);
		ret+=msgLen;
		std::vector<byte> buffVec(buffArray.get(), buffArray.get() + msgLen);

		BinaryProtocol proto;
		proto.setBuffer(buffVec);
		ResponseMsg msg;
		msg.deSerialize(&proto);
		msgVec.push_back(msg);
	}

	return ret;
}

int rpc::TcpTransport::sendResponseMsg(const ResponseMsg& respMsg) {
	BinaryProtocol proto;
	respMsg.serialize(&proto);

	uint16 len = uint16(proto.getBuffer().size());
	struct evbuffer * msgbuff= evbuffer_new();
	evbuffer_add(msgbuff,(byte*)&len,MsgHeaderMaxSize);
	evbuffer_add(msgbuff,proto.getBuffer().data(), proto.getBuffer().size());

	int ret = bufferevent_write_buffer(bev_, msgbuff);

	evbuffer_free(msgbuff);

	return ret;
}

int rpc::TcpTransport::recvRequestMsg(struct evbuffer* buff,std::vector<RequestMsg>& msgVec) {
	int ret=0;
	while(1){
		if (evbuffer_get_length(buff) <= MsgHeaderMaxSize) {
			return ret;
		}
		struct evbuffer_iovec v[1];
		ret=evbuffer_peek(buff, MsgHeaderMaxSize, NULL, v, 1);
		RpcAssert(ret==1);
		uint16 msgLen= *(uint16*)(v[0].iov_base);

		if (evbuffer_get_length(buff) < size_t( msgLen+MsgHeaderMaxSize)) {
			return ret;
		}
		evbuffer_drain(buff,MsgHeaderMaxSize);
		ret+=MsgHeaderMaxSize;

		std::unique_ptr<byte[]> buffArray = std::make_unique<byte[]>(msgLen);
		evbuffer_remove(buff, buffArray.get(), msgLen);
		ret+=msgLen;
		std::vector<byte> buffVec(buffArray.get(), buffArray.get() + msgLen);

		BinaryProtocol proto;
		proto.setBuffer(buffVec);
		RequestMsg msg;
		msg.deSerialize(&proto);
		msgVec.push_back(msg);
	}

	return ret;
}

File: lib/go/rpc/Transport.go
package rpc

import (
	"bytes"
	"encoding/binary"
	"io"
)

const MsgHeaderMaxSize = 2

type ITransport interface {
	sendRequestMsg(reqMsg *RequestMsg) error
	recvResponseMsg(buff *bytes.Buffer) (*ResponseMsg, error)
	sendResponseMsg(respMsg *ResponseMsg) error
	recvRequestMsg(buff *bytes.Buffer) (*RequestMsg, error)
	start(conn *connection) error
}
type ITransportFactory interface {
	NewTransport() ITransport
}

type TcpTransportFactory struct {
}

func (this TcpTransportFactory) NewTransport() ITransport {
	return &TcpTransport{}
}

type TcpTransport struct {
	*connection
}

func (this *TcpTransport) sendRequestMsg(reqMsg *RequestMsg) error {
	proto := &BinaryProtocol{buff: new(bytes.Buffer)}
	reqMsg.Serialize(proto)
	this.SendBuffer(proto.GetBuffer())
	return nil
}
func (this *TcpTransport) recvResponseMsg(buff *bytes.Buffer) (*ResponseMsg, error) {
	proto := &BinaryProtocol{}
	proto.SetBuffer(buff)
	var msg ResponseMsg
	msg.DeSerialize(proto)
	return &msg, nil

}

func (this *TcpTransport) sendResponseMsg(respMsg *ResponseMsg) error {
	proto := &BinaryProtocol{buff: new(bytes.Buffer)}
	respMsg.Serialize(proto)
	this.SendBuffer(proto.GetBuffer())
	return nil
}
func (this *TcpTransport) recvRequestMsg(buff *bytes.Buffer) (*RequestMsg, error) {
	proto := &BinaryProtocol{}
	proto.SetBuffer(buff)
	var msg RequestMsg
	msg.DeSerialize(proto)
	return &msg, nil
}
func (this *TcpTransport) start(conn *connection) error {
	this.connection = conn

	this.conn.SetNoDelay(true)
	this.conn.SetKeepAlive(true)

	//同步退出 goroutine
	this.exitSync.Add(2)

	//开启读写 goroutine
	go this.recvMsgLoop()
	go this.sendMsgLoop()

	return nil
}

func (this *TcpTransport) SendBuffer(buff *bytes.Buffer) bool {
	//推入发送循环
	this.sendCh <- buff
	return true
}

func (this *TcpTransport) sendMsgLoop() {
	for msg := range this.sendCh {
		if msg == nil || this.conn == nil {
			break
		}

		msgLen := uint16(msg.Len())
		buf := make([]byte, MsgHeaderMaxSize)
		binary.LittleEndian.PutUint16(buf, msgLen)
		_, err := this.conn.Write(buf)
		if err != nil {
			break
		}
		_, err = this.conn.Write(msg.Bytes())
		if err != nil {
			break
		}

	}
	//关闭socket 从读操作退出
	this.shutDown()
	this.exitSync.Done()
}

func (this *TcpTransport) recvMsgLoop() {
	var err error
	var msgLen int
	msgHeader := make([]byte, MsgHeaderMaxSize)
	for {
		_, err = io.ReadFull(this.conn, msgHeader)
		if err != nil {
			break
		}
		msgLen = int(binary.LittleEndian.Uint16(msgHeader))
		if msgLen <= 0 || msgLen > 65535 {
			break
		}

		msgBody := make([]byte, msgLen)
		_, err = io.ReadFull(this.conn, msgBody)
		if err != nil {
			break
		}
		cMsg := NewCommonMsg(this.cb, 1, bytes.NewBuffer(msgBody))
		this.event.queue <- cMsg
	}
	this.exitSync.Done()
}

func (this *TcpTransport) shutDown() {
	// 关闭连接
	if this.conn != nil {
		this.conn.Close()
		this.conn = nil
	}
}
